### DISCUSSION TITLE
[Bug]: watch() doesn't return defaultValues after rerender with new defaultValues

### DISCUSSION THREAD

**Discussion by t13m** [react-hook-form/react-hook-form#6302] [Q&A]
### Version Number

7.12.2

### Codesandbox/Expo snack

https://codesandbox.io/s/rhf-watch-reset-issue-demo-go2d4?file=/src/App.tsx

### Steps to reproduce

1. Go to https://codesandbox.io/s/rhf-watch-reset-issue-demo-go2d4?file=/src/App.tsx
2. Click on 'Switch to ObjB' button
3. The App will crash. Logs in the console show that the prop has new value but the value returned by watch() is not consistent with prop (or with defaultValues fed to useForm).

### Expected behaviour

The scenario is to build a cascaded select box, the option values in it are looked up in the props with previous selected value as key. When user switch to other object, the values should change accordingly.

Also, I don't know if the usage of reset for this kind of scenario is a good practice (The form will render twice because of the useEffect and reset). Maybe a complete unmounting and then mounting the form with new props (by changing the `key` of  component) would be a better solution?

### What browsers are you seeing the problem on?

Chrome, Safari

### Relevant log output

```shell
Current tags from props: B1,B2 
Current selected tag from watch(): A1
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
After you invoke the reset function, `defaultValues` will be updated into the values which you have supplied.

**Reply by bluebill1049** [reply]
```tsx
const defaultValues = { a: 'test' }

useForm({
  defaultValues
})

rest({
  something: 'else' // this will be the new `defaultValues` and it's best to avoid the same variable assignment to avoid mutation.
})
```

**Reply by t13m** [reply]
Thank you so much for reply so quickly!

As you mentioned that it's best to avoid same variable assignment to avoid mutation, and the fact that values are updated with defaultValues after `reset`. I think maybe the correct way to handle this scenario (form values always be the properties of current focused object which may be changed) should be completely re-mounting the component, but  not `useEffect( () => { reset(newValue) }, [newValue] )` way. What do you think? Thanks for your time again :) And this library is truly an excellent work, I like it.

**Reply by bluebill1049** [reply]
no worries, is there any way to avoid the `reset` for your use case? maybe doing the switch data at your component level? and trigger the switch by watch/useWatch?

**Reply by t13m** [reply]
`reset` is used for switching between different object (reference: https://stackoverflow.com/a/66268317 ).

In my case, there are several geometry objects displayed in a canvas, while properties of the one currently selected by user will be showed in an inspector, which is made with RHF. Basically, when user select an object or modify its size (on canvas), the values of the form should change accordingly.

Sorry but what do you mean by "trigger the switch by watch/useWatch"?

Don't think `reset` can be avoided because it is used to change form values according to changes of external states, but yes, the switching of data can be done by re-mounting the component. This method is ok for me now.

**Reply by bluebill1049** [reply]
i was meant for `setValue` and then you can `watch` that update without re-mount the entire form, I could be wrong here tho.
